subsys/net/lib/openthread/platform/uart.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2020 Tridonic GmbH & Co KG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_LEVEL CONFIG_OPENTHREAD_LOG_LEVEL
+#define LOG_MODULE_NAME net_otPlat_uart
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <kernel.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <drivers/uart.h>
+
+#include <sys/ring_buffer.h>
+#include <sys/atomic.h>
+
+#ifdef CONFIG_OPENTHREAD_NCP_SPINEL_ON_UART_ACM
+#include <usb/usb_device.h>
+#endif
+
+#include <openthread-system.h>
+#include <openthread/platform/uart.h>
+
+#include "platform-zephyr.h"
+
+struct openthread_uart {
+	struct ring_buf *tx_ringbuf;
+	struct ring_buf *rx_ringbuf;
+	struct device *dev;
+	atomic_t tx_busy;
+	atomic_t tx_finished;
+};
+
+#define OT_UART_DEFINE(_name, _ringbuf_size) \
+	RING_BUF_DECLARE(_name##_tx_ringbuf, _ringbuf_size); \
+	RING_BUF_DECLARE(_name##_rx_ringbuf, _ringbuf_size); \
+	static struct openthread_uart _name = { \
+		.tx_ringbuf = &_name##_tx_ringbuf, \
+		.rx_ringbuf = &_name##_rx_ringbuf, \
+	};
+
+OT_UART_DEFINE(ot_uart, CONFIG_OPENTHREAD_NCP_UART_RING_BUFFER_SIZE)
+
+#define RX_FIFO_SIZE 128
+
+static void uart_rx_handle(/*const struct openthread_uart *ot_uart*/)
+{
+	u8_t *data;
+	u32_t len;
+	u32_t rd_len;
+	bool new_data = false;
+
+	do {
+		len = ring_buf_put_claim(
+			ot_uart.rx_ringbuf, &data,
+			ot_uart.rx_ringbuf->size);
+		if (len > 0) {
+			rd_len = uart_fifo_read(
+				ot_uart.dev, data, len);
+
+			if (rd_len > 0) {
+				new_data = true;
+			}
+
+			int err = ring_buf_put_finish(
+				ot_uart.rx_ringbuf, rd_len);
+			(void)err;
+			__ASSERT_NO_MSG(err == 0);
+		} else {
+			u8_t dummy;
+
+			/* No space in the ring buffer - consume byte. */
+			LOG_WRN("RX ring buffer full.");
+
+			rd_len = uart_fifo_read(
+				ot_uart.dev, &dummy, 1);
+		}
+	} while (rd_len && (rd_len == len));
+
+	if (new_data) {
+		otSysEventSignalPending();
+	}
+}
+
+static void uart_tx_handle(/*struct openthread_uart *ot_uart*/)
+{
+	u32_t len;
+
+	const u8_t *data;
+
+	len = ring_buf_get_claim(ot_uart.tx_ringbuf, (u8_t **)&data,
+				 ot_uart.tx_ringbuf->size);
+	if (len) {
+		int err;
+
+		len = uart_fifo_fill(ot_uart.dev, data, len);
+		err = ring_buf_get_finish(ot_uart.tx_ringbuf, len);
+		(void)err;
+		__ASSERT_NO_MSG(err == 0);
+	} else {
+		uart_irq_tx_disable(ot_uart.dev);
+		ot_uart.tx_busy = 0;
+		atomic_set(&(ot_uart.tx_finished), 1);
+		otSysEventSignalPending();
+	}
+}
+
+static void uart_callback(void *user_data)
+{
+	while (uart_irq_update(ot_uart.dev) &&
+	       uart_irq_is_pending(ot_uart.dev)) {
+
+		if (uart_irq_rx_ready(ot_uart.dev)) {
+			uart_rx_handle();
+		}
+
+		if (uart_irq_tx_ready(ot_uart.dev)) {
+			uart_tx_handle();
+		}
+	}
+};
+
+void platformUartProcess(otInstance *aInstance)
+{
+	u32_t len = 0;
+	const u8_t *data;
+
+	/* Process UART RX */
+	while (
+				(
```

Comment:
This indentation does not seem right.

Reply:
fixed.

Reply:
Not fixed as well. Did you miss a push?

Just make it something like:
```
	while ((len = ring_buf_get_claim(ot_uart.rx_ringbuf,
					 (u8_t **)&data,
					 ot_uart.rx_ringbuf->size) > 0) {
```